src/support_tests.cc

```diff
@@ -767,6 +770,135 @@ bool test_digest(bool print_all) {
   return true;
 }
 
+#define NUM_TEST_FILES_TO_DIGEST 4
+
+/*
+ * test_digest_multiple() - Similar to test_digest()
+ *
+ * Here we do a simple verification whether the hash-digest produced by
+ * digest_message() is the same if you read-in all input files once
+ * v/s if you read-on each file one-at-a-time and make multiple calls
+ * to digest_message(). Using digest_message(), which is a wrapper around
+ * underlying EVP_Digest*() interfaces, is not additive. This test checks
+ * that the resultant digest is different between the two schemes.
+ */
+bool test_digest_multiple(bool print_all) {
+  // This test-case will be invoked by ./certifier_tests.exe from the src/
+  // dir. Create a list of files to build the hash digest for verification.
+  const char *files_list[] = {"./certifier_tests.exe",
+                              "./support_tests.cc",
+                              "../libcertifier_framework.so",
+                              "../certifier_framework.py"};
+  int         files_size[NUM_TEST_FILES_TO_DIGEST] = {0};
+
+  // Read-in other files to capture each file's size
+  int         max_file_size = 0;
+  int         all_files_total_size = 0;
+  const char *filename = nullptr;
+  for (int i = 0; i < NUM_TEST_FILES_TO_DIGEST; i++) {
+    filename = files_list[i];
+    int this_files_size = file_size(filename);
+    if (this_files_size < 0) {
+      printf("Error: Input file '%s' not found.\n", filename);
+      return false;
+    }
+    files_size[i] = this_files_size;
+    if (max_file_size < this_files_size) {
+      max_file_size = this_files_size;
+    }
+    all_files_total_size += this_files_size;
+  }
+
+  byte *to_hash_single = (byte *)malloc(max_file_size);
+  if (to_hash_single == nullptr) {
+    printf("Error: Cannot malloc %d bytes for to_hash_single[]\n",
+           max_file_size);
+    return false;
+  }
+
+  const int sha256_size = 32;
+
+  byte digest_updated[sha256_size];
+  memset(digest_updated, 0, sizeof(digest_updated));
+
+  // Compute the aggregated hash digest, one file-at-a-time.
+  for (int i = 0; i < NUM_TEST_FILES_TO_DIGEST; i++) {
+    filename = files_list[i];
+    if (!read_file(filename, &files_size[i], to_hash_single)) {
+      free(to_hash_single);
+      printf("Error: Cannot read file '%s'\n", filename);
+      return false;
+    }
+    if (print_all) {
+      printf("Read file: '%s', size=%d\n", filename, files_size[i]);
+    }
+    // Update digest-hash across each file
+    if (!digest_message(Digest_method_sha256,
```

Comment:
I noted that this wrapper interface is calling stuff like `EVP_DigestUpdate()`, followed by `EVP_DigestFinal_ex()`.

Seems like once you finalize the digest, then you have to re-start on the next call to this `digest_message()` with a new `EVP_MD_CTX *mdctx`.

I feel unless we re-work this `digest_message()` to -not- call finalize, so that it can be called iteratively on multiple files, we won't get the same resultant hash for all-files-in-one-go v/s each-file-digest method.

Let me know if I'm seeing things wrong ... or if this test-case is bogus and misadvertised.

Reply:
Correct. The current implementation of `digest_message` doesn't allow continuous hashing of multiple inputs.